traffic_ops/traffic_ops_golang/steering/steering.go

```diff
@@ -0,0 +1,233 @@
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import (
+	"database/sql"
+	"errors"
+	"net/http"
+
+	"github.com/apache/trafficcontrol/lib/go-tc"
+	"github.com/apache/trafficcontrol/lib/go-util"
+	"github.com/apache/trafficcontrol/traffic_ops/traffic_ops_golang/api"
+
+	"github.com/lib/pq"
+)
+
+func Get(w http.ResponseWriter, r *http.Request) {
+	inf, userErr, sysErr, errCode := api.NewInfo(r, nil, nil)
+	if userErr != nil || sysErr != nil {
+		api.HandleErr(w, r, inf.Tx.Tx, errCode, userErr, sysErr)
+		return
+	}
+	defer inf.Close()
+
+	steering, err := findSteering(inf.Tx.Tx)
+	if err != nil {
+		api.HandleErr(w, r, inf.Tx.Tx, http.StatusInternalServerError, nil, errors.New("steering.Get finding: "+err.Error()))
+		return
+	}
+	api.WriteResp(w, r, steering)
+}
+
+func findSteering(tx *sql.Tx) ([]tc.Steering, error) {
+	steeringData, err := getSteeringData(tx)
+	if err != nil {
+		return nil, err
+	}
+	targetIDs := steeringDataTargetIDs(steeringData)
+	steeringFilters, err := getSteeringFilters(tx, targetIDs)
+	if err != nil {
+		return nil, err
+	}
+	primaryOriginCoords, err := getPrimaryOriginCoords(tx, targetIDs)
+	if err != nil {
+		return nil, err
+	}
+
+	steerings := map[tc.DeliveryServiceName]tc.Steering{}
+
+	for _, data := range steeringData {
+		if _, ok := steerings[data.DeliveryService]; !ok {
+			steerings[data.DeliveryService] = tc.Steering{
+				DeliveryService: data.DeliveryService,
+				ClientSteering:  data.DSType == tc.DSTypeClientSteering,
+				Filters:         []tc.SteeringFilter{},         // Initialize, so JSON produces `[]` not `null` if there are no filters.
+				Targets:         []tc.SteeringSteeringTarget{}, // Initialize, so JSON produces `[]` not `null` if there are no targets.
+			}
+		}
+		steering := steerings[data.DeliveryService]
+
+		if filters, ok := steeringFilters[data.TargetID]; ok {
+			steering.Filters = append(steering.Filters, filters...)
+		}
+
+		target := tc.SteeringSteeringTarget{DeliveryService: data.TargetName}
+		switch data.Type {
+		case tc.SteeringTypeOrder:
+			target.Order = uint64(data.Value)
+		case tc.SteeringTypeWeight:
+			target.Weight = uint64(data.Value)
+		case tc.SteeringTypeGeoOrder:
+			target.GeoOrder = util.IntPtr(data.Value)
+			target.Latitude = util.FloatPtr(primaryOriginCoords[data.TargetID].Lat)
```

Comment:
Just for clarification, if the primary origin doesn't have a coordinate assigned, this will still set target.Latitude and Longitude to 0.0? I _think_ that's what happens with this map when `data.TargetID` isn't a key in the map.

Reply:
Yes, if the key isn't in the map because the origin had no coordinate, Go will return the default value of `Coord` i.e. `{0.0,0.0}`. Which looks like what Perl is doing, unless I'm misreading it?